src/prefect/orion/database/orm_models.py

```diff
@@ -187,7 +216,36 @@ def task_run_id(cls):
         default=schemas.states.StateDetails,
         nullable=False,
     )
-    data = sa.Column(sa.JSON, nullable=True)
+    _data = sa.Column(sa.JSON, nullable=True)
+
+    @declared_attr
+    def result_artifact_id(cls):
+        return sa.Column(
+            UUID(),
+            sa.ForeignKey(
+                "artifact.id",
+                ondelete="SET NULL",
+                use_alter=True,
+            ),
+            index=True,
+        )
+
+    @declared_attr
+    def _result_artifact(cls):
+        return sa.orm.relationship(
+            "Artifact",
+            lazy="joined",
+            foreign_keys=[cls.result_artifact_id],
+            primaryjoin="Artifact.id==%s.result_artifact_id" % cls.__name__,
+        )
+
+    @hybrid_property
+    def data(self):
+        if self._data:
+            return self._data
+        if not self.result_artifact_id:
+            return None
+        return getattr(self._result_artifact, "data", None)
```

Comment:
When can this attribute be missing given that `data` is a field on `ORMArtifact`?

Reply:
It shouldn't be, this is leftover from an earlier implementation, so I'm removing the `getattr` call